packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -560,115 +781,160 @@ mixin DirectionalFocusTraversalPolicyMixin on FocusTraversalPolicy {
-  /// Creates a const [WidgetOrderFocusTraversalPolicy].
-  WidgetOrderFocusTraversalPolicy();
-
+///  * [OrderedTraversalPolicy], a policy that describes the order
+///    explicitly using [FocusTraversalOrder] widgets.
+class WidgetOrderTraversalPolicy extends FocusTraversalPolicy with DirectionalFocusTraversalPolicyMixin {
   @override
-  FocusNode findFirstFocus(FocusNode currentNode) {
-    assert(currentNode != null);
-    final FocusScopeNode scope = currentNode.nearestScope;
-    // Start with the candidate focus as the focused child of this scope, if
-    // there is one. Otherwise start with this node itself. Keep going down
-    // through scopes until an ultimately focusable item is found, a scope
-    // doesn't have a focusedChild, or a non-scope is encountered.
-    FocusNode candidate = scope.focusedChild;
-    if (candidate == null) {
-      if (scope.traversalChildren.isNotEmpty) {
-        candidate = scope.traversalChildren.first;
-      } else {
-        candidate = currentNode;
-      }
-    }
-    while (candidate is FocusScopeNode && candidate.focusedChild != null) {
-      final FocusScopeNode candidateScope = candidate as FocusScopeNode;
-      candidate = candidateScope.focusedChild;
-    }
-    return candidate;
+  Iterable<FocusNode> sortDescendants(Iterable<FocusNode> descendants) => descendants;
+}
+
+// This class exists mainly for efficiency reasons: the rect is copied out of
+// the node, because it will be accessed many times in the reading order
+// algorithm, and the FocusNode.rect accessor does coordinate transformation. If
+// not for this optimization, it could just be removed, and the node used
+// directly.
+//
+// It's also a convenient place to put some utility functions having to do with
+// the sort data.
+class _ReadingOrderSortData extends Diagnosticable {
+  _ReadingOrderSortData(this.node)
+      : assert(node != null),
+        rect = node.rect,
+        directionality = _findDirectionality(node.context);
+
+  final TextDirection directionality;
+  final Rect rect;
+  final FocusNode node;
+
+  // Find the directionality in force for a build context without creating a
+  // dependency.
+  static TextDirection _findDirectionality(BuildContext context) {
+    return (context.getElementForInheritedWidgetOfExactType<Directionality>()?.widget as Directionality)?.textDirection;
   }
 
-  // Moves the focus to the next or previous node, depending on whether forward
-  // is true or not.
-  bool _move(FocusNode currentNode, {@required bool forward}) {
-    if (currentNode == null) {
-      return false;
+  /// Finds the common Directional ancestor of an entire list of groups.
+  static TextDirection commonDirectionalityOf(List<_ReadingOrderSortData> list) {
+    final List<List<Directionality>> allAncestors = list.map<List<Directionality>>((_ReadingOrderSortData member) => member.directionalAncestors.toList()).toList();
```

Comment:
Could this just be an `Iterable<Set<Directionality>` to avoid one of the toList() calls on this line and the toSet call a few lines below?